go/obscuronode/enclave/sql/edgelessdb.go

```diff
@@ -0,0 +1,480 @@
+	}
+	caBytes, err := x509.CreateCertificate(rand.Reader, caCert, caCert, &caPrivKey.PublicKey, caPrivKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create CA cert - %w", err)
+	}
+	caPEMBuf := new(bytes.Buffer)
+	err = pem.Encode(caPEMBuf, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: caBytes,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create ca cert pem - %w", err)
+	}
+
+	caPEM := caPEMBuf.String()
+	manifest := &manifest{
+		SQL:  manifestSQLStatements,
+		Cert: caPEM,
+		// IF DEBUGGING: this should be set to enable verbose logging
+		//     Note: it also requires EDG_EDB_DEBUG=1 on the process, see https://docs.edgeless.systems/edgelessdb/#/reference/configuration
+		// Debug: true,
+	}
+	err = initialiseEdgelessDB(edbHost, manifest, edbPEM)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialise edgeless DB with created manifest - %w \nmanifest: %v", err, manifest)
+	}
+
+	// store certificates for DB connection
+	err = prepareCertificates(caCert, caPEM, caPrivKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare and persist certs for edb connection - %w", err)
+	}
+
+	// persist the manifest for any future restarts of the enclave
+	err = writeManifest(manifest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to persist manifest file - %w", err)
+	}
+
+	return manifest, nil
+}
+
+// initialiseEdgelessDB sends a manifest over http to the edgeless DB with its initial config
+func initialiseEdgelessDB(edbHost string, m *manifest, edbPEM string) error {
+	b, err := json.Marshal(m)
+	if err != nil {
+		return fmt.Errorf("failed to marshal manifest json - %w", err)
+	}
+	url := fmt.Sprintf("https://%s:%s/manifest", edbHost, edbHTTPPort)
+	req, err := http.NewRequestWithContext(context.Background(), "POST", url, bytes.NewBuffer(b))
+	if err != nil {
+		return fmt.Errorf("faild to create manifest initialization req - %w", err)
+	}
+	caCertPool := x509.NewCertPool()
+	if ok := caCertPool.AppendCertsFromPEM([]byte(edbPEM)); !ok {
+		return fmt.Errorf("failed to append to CA cert from edb cert pem")
+	}
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs:    caCertPool,
+				MinVersion: tls.VersionTLS12,
+			},
+		},
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("manifest initialization req failed - %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode > 299 {
+		var msg []byte
+		_, err := resp.Body.Read(msg)
+		if err != nil {
+			return fmt.Errorf("manifest initialization req failed with status code: %d, failed to read status text", resp.StatusCode)
+		}
+		return fmt.Errorf("manifest initialization req failed with status: %d %s", resp.StatusCode, msg)
+	}
+	return nil
+}
+
+// verifyEdgelessDB requests the /signature from the edb, it should match the hash of the manifest we expected
+func verifyEdgelessDB(edbHost string, m *manifest, edbPEM string) error {
```

Comment:
@BedrockSquirrel The first 80% of this method is the logic in `initialiseEdgelessDB`. Can you try pulling it out into a helper method?

Reply:
Think this is helped with just creating the http client once hopefully